{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "import sys\n",
    "import urllib.parse\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pdf_url(doi, base_url):\n",
    "    url = f\"{base_url}/{doi}\"\n",
    "    headers = {\n",
    "        \"User-Agent\": \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3\"\n",
    "    }\n",
    "\n",
    "    try:\n",
    "        response = requests.get(url, headers=headers, timeout=30)\n",
    "        response.raise_for_status()\n",
    "\n",
    "        soup = BeautifulSoup(response.content, \"html.parser\")\n",
    "\n",
    "        # Find direct PDF link\n",
    "        iframe = soup.find(\"iframe\", id=\"pdf\")\n",
    "        if iframe:\n",
    "            pdf_path = iframe.get(\"src\")\n",
    "            if pdf_path:\n",
    "                return urllib.parse.urljoin(base_url, pdf_path)\n",
    "\n",
    "        # Alternative for newer Sci-Hub versions\n",
    "        button = soup.find(\"button\", onclick=lambda x: x and \"location.href\" in x)\n",
    "        if button:\n",
    "            js_code = button[\"onclick\"]\n",
    "            pdf_path = js_code.split(\"=\")[1].strip(\"';\")\n",
    "            return urllib.parse.urljoin(base_url, pdf_path)\n",
    "\n",
    "        return None\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error accessing {base_url}: {str(e)}\")\n",
    "        return None\n",
    "\n",
    "\n",
    "def download_pdf(pdf_url, filename, folder_path):\n",
    "    headers = {\"User-Agent\": \"Mozilla/5.0\"}\n",
    "    try:\n",
    "        response = requests.get(pdf_url, headers=headers, stream=True)\n",
    "        response.raise_for_status()\n",
    "        \n",
    "        # Create the folder if it doesn't exist\n",
    "        if not os.path.exists(folder_path):\n",
    "            os.makedirs(folder_path)\n",
    "\n",
    "        # Create the file in the folder\n",
    "        with open(os.path.join(folder_path, filename), \"wb\") as f:\n",
    "            for chunk in response.iter_content(chunk_size=8192):\n",
    "                if chunk:\n",
    "                    f.write(chunk)\n",
    "        return True\n",
    "    except Exception as e:\n",
    "        print(f\"Error during download: {str(e)}\")\n",
    "        return False\n",
    "    \n",
    "def get_doi_from_doi_url(doi_url):\n",
    "    # Get the DOI from the url \"https://doi.org/10.1149/1.2069372\" -> \"10.1149/1.2069372\"\n",
    "    doi = doi_url.split(\"https://doi.org/\")[1]\n",
    "    return doi\n",
    "\n",
    "\n",
    "def get_pdf_from_doi_url(doi_url, folder_path, ID):\n",
    "    if len(sys.argv) != 2:\n",
    "        print(\"Usage: python scihub_downloader.py <DOI>\")\n",
    "        sys.exit(1)\n",
    "\n",
    "    scihub_domains = [\n",
    "        \"https://sci-hub.se\",\n",
    "        \"https://sci-hub.st\",\n",
    "        \"https://sci-hub.ru\",\n",
    "        \"https://sci-hub.ee\",\n",
    "    ]\n",
    "\n",
    "    pdf_url = None\n",
    "    for domain in scihub_domains:\n",
    "        print(f\"Trying with domain: {domain}\")\n",
    "        doi = get_doi_from_doi_url(doi_url)\n",
    "        pdf_url = get_pdf_url(doi, domain)\n",
    "        if pdf_url:\n",
    "            print(f\"PDF found on {domain}\")\n",
    "            break\n",
    "\n",
    "    if not pdf_url:\n",
    "        print(\"Failed to locate PDF on all Sci-Hub domains\")\n",
    "        return\n",
    "    filename = ID + \".pdf\"\n",
    "    \n",
    "    print(f\"Attempting download from: {pdf_url}\")\n",
    "\n",
    "    if download_pdf(pdf_url, filename, folder_path):\n",
    "        print(f\"PDF successfully saved as: {filename}\")\n",
    "    else:\n",
    "        print(\"Download failed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDF déjà téléchargé pour D. Guyomard and J.M. Tarascon, J. Electrochem. Soc. Vol 175 (1992)\n",
      "Téléchargement du PDF pour A. Blyr, C. Sigala et al. Jr. Electrochemical Soc. Vol. 145, 1998\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.1149/1.1838235 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x113557d90>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.1149/1.1838235 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ac690>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.1149/1.1838235 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142aca50>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.1149/1.1838235 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ad090>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "PDF déjà téléchargé pour P. Poizot et al. Nature, 407, 496-499, 2000\n",
      "PDF déjà téléchargé pour X. Ma... et al. Journal of the electrochemical society, 166(4) 2019\n",
      "PDF déjà téléchargé pour K.J. Nelson et al. Journal of the electrochemical society, 165, 2018\n",
      "PDF déjà téléchargé pour G.P. Aiken et al. Journal of he electrochemical society. 167 (2020).\n",
      "PDF déjà téléchargé pour J.E. Harlow et al. J. Electrochemical Society, 165 (2018)\n",
      "PDF déjà téléchargé pour H.M. Dahn et al. Journal of the electrochemical society, 159(9) 2012\n",
      "PDF déjà téléchargé pour K.J. Nelson et al. et al. Journal of the electrochemical society, 165(3) 2018\n",
      "PDF déjà téléchargé pour G. Yan et al. Journal of the electrochemical society, 165(2) 2018\n",
      "PDF déjà téléchargé pour M. Dollé et al. Journal of the electrochemical society A851-A857\n",
      "PDF déjà téléchargé pour J. Li, H. Liu, J. Xia et al, Journal of the Electrochemical Society, 164 (4) 2017\n",
      "PDF déjà téléchargé pour R.P. Day et al. Journal of the Electrochemical society, 162(14) 2015\n",
      "PDF déjà téléchargé pour D.P. Wilkinson, H. Blom, K. Brandt and D. Wainwright, Jr of power sources, 36 (1991)\n",
      "PDF déjà téléchargé pour Glazier et al J. Electrochem Soc. A3545-55 (2017)\n",
      "PDF déjà téléchargé pour Joule 3 1-17 March 20 2019\n",
      "PDF déjà téléchargé pour X. Wang et al. JECS. 154 (2007)\n",
      "PDF déjà téléchargé pour Knobloch et al. Journal of electronic packaging, Vol. 140 (2018)\n",
      "PDF déjà téléchargé pour J. Bonefacino. Thesis. Department of Electrical Engineering. The Hong Kong Polytechnic University. (2017)\n",
      "Téléchargement du PDF pour F Udd and WB Spillman Fiber optic sensors 2011\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.1002/9781118014103.ch7 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ad590>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.1002/9781118014103.ch7 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ada90>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.1002/9781118014103.ch7 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142adf90>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.1002/9781118014103.ch7 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x113557890>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "PDF déjà téléchargé pour J. Huang L Albero Blanquer et al. Nature energy August 2020\n",
      "PDF déjà téléchargé pour E Hud Proc of SPIE Vol 9098 (2014)\n",
      "PDF déjà téléchargé pour J. Bonefacino Light: Science & Applications 7 (3). (2018)\n",
      "PDF déjà téléchargé pour J. Bonefacino Light: Science & Applications 7 (3). (2018)\n",
      "PDF déjà téléchargé pour D. Vilarinho Sensors 2017, 17. 2924; doi:10.3390/s17122924\n",
      "PDF déjà téléchargé pour G. Yang et al. Measurement 46 (2013)\n",
      "Téléchargement du PDF pour M. Nascimento et al. Applied Thermal Engineering 149 (2019)\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.1016/j.energy.2023.127479 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142add10>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.1016/j.energy.2023.127479 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ad810>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.1016/j.energy.2023.127479 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ad310>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.1016/j.energy.2023.127479 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ace10>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "PDF déjà téléchargé pour LW. Sommer et al: JECS 162 (14) 2015.\n",
      "PDF déjà téléchargé pour L.W. Sommer et al. Mater. Res. Soc. Symp. Proc. Vol. 1681 @ 2014\n",
      "Téléchargement du PDF pour LW. Sommer et al; JECS 162 (14) 2015.\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.1186/s12887-021-03005-3 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ac550>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.1186/s12887-021-03005-3 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ac410>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.1186/s12887-021-03005-3 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ae490>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.1186/s12887-021-03005-3 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ae990>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "PDF déjà téléchargé pour J. Schwartz, Sommer et al. Mater. Res. Soc. Symp. Proc. Vol. 1740 © 2015\n",
      "PDF déjà téléchargé pour CJ Bae et al. Energy technology, 4, 2016\n",
      "PDF déjà téléchargé pour R. Gafsiand M.A. El-Sherif, Optical fiber technology 6, 299-323 (2000)\n",
      "PDF déjà téléchargé pour R.B Wagreich et al. Electronics letters June 1996 Vol. 32, N° 13\n",
      "PDF déjà téléchargé pour R. Gafsiand M.A. El-Sherif, Optical fiber technology 6, 299-323 (2000)\n",
      "PDF déjà téléchargé pour M. Nascimento et al. Batteries 2018, 4, 67\n",
      "Téléchargement du PDF pour Pinto et al. Journal of Power Sources 1-9, 2018\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.1186/s13705-024-00475-y (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142aee90>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.1186/s13705-024-00475-y (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142af390>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.1186/s13705-024-00475-y (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142af890>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.1186/s13705-024-00475-y (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142afd90>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "PDF déjà téléchargé pour J. Flemming et al. HardwareX 3 (2018)\n",
      "PDF déjà téléchargé pour J. Huang, L. Albero Blanquer et al, Nature energy August 2020\n",
      "PDF déjà téléchargé pour J. Huang, L. Albero Blanquer et al, Nature energy August 2020\n",
      "PDF déjà téléchargé pour J. Huang, L. Albero Blanquer et al., Nature Energy, 2020, 5, 674-683\n",
      "PDF déjà téléchargé pour F. Chiavaioli et al. Nanophotoncs 2017.\n",
      "PDF déjà téléchargé pour M. Morisawa et al. Jornal of sensors 2012\n",
      "PDF déjà téléchargé pour Tai, et al. Opt. Lett. 1987\n",
      "PDF déjà téléchargé pour Tai. et al. Opt. Lett. 1987\n",
      "PDF déjà téléchargé pour Nieva, et al. ACS AMI. 2016\n",
      "PDF déjà téléchargé pour P. Nieva et al. ACS applied materials and interfaces, 2017, 9, 41284, 9,\n",
      "PDF déjà téléchargé pour J. Albert et al., Laser & Photonics, 2013, 7.83-108.\n",
      "PDF déjà téléchargé pour C. Caucheteur et al. Nature Communications 2016.\n",
      "PDF déjà téléchargé pour T. Guo et al., Optics and Laser Technology, 2016, 78, 19-33.\n",
      "PDF déjà téléchargé pour T. Guao, H.W Tam et al Optics Expres Vol 17, 2009\n",
      "PDF déjà téléchargé pour A. Nedialkov et al. Batteries march 2019\n",
      "PDF déjà téléchargé pour L. Marec, L. Quetel, et al. Insitu optical fibre sensors for temperature and salinity monitoring, OCS 2005\n",
      "PDF déjà téléchargé pour J.H. Chong et al. Optics Communications 229 (2004)\n",
      "PDF déjà téléchargé pour Marrec, L., Bourgerette, T., Datin, E., Ferchaud, N., Pucel, B., Quétel, L., Renault, C., & Tregoat, D. (2005). In-situ optical fibre sensors for temperature and salinity monitoring. In Oceans 2005 - Europe (Vol. 2, pp. 1276-1278). IEEE\n",
      "Téléchargement du PDF pour J. Janting, Journal of lightwave technology\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.1364/ofs.2018.tue63 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142702d0>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.1364/ofs.2018.tue63 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142707d0>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.1364/ofs.2018.tue63 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x114270cd0>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.1364/ofs.2018.tue63 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x114287610>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "Téléchargement du PDF pour Small and Robust All-Polymer Fiber Bragg Grating Based pH Sensor, DOI 10.1109/JLT. 2019.2902638\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.1364/ofs.2018.the52 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x113557c50>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.1364/ofs.2018.the52 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142afb10>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.1364/ofs.2018.the52 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142af610>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.1364/ofs.2018.the52 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142af110>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "PDF déjà téléchargé pour A. Trouillet et al. Meas. Science Technology 17 (2006)\n",
      "PDF déjà téléchargé pour J. Dai et al, Sensors and Actuators B. 190 (2014)\n",
      "Téléchargement du PDF pour A. M. Cubillas et al. Chem Soc. Rev 2013\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.1364/lacsea.2014.lm4d.5 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142aec10>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.1364/lacsea.2014.lm4d.5 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ae710>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.1364/lacsea.2014.lm4d.5 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ae210>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.1364/lacsea.2014.lm4d.5 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ac690>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "Téléchargement du PDF pour S Maquis, G. Laffont, P. Ferdinand, et al. Optics Express, 2008\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.4324/9781315608730 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142aca50>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.4324/9781315608730 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ad090>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.4324/9781315608730 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ad590>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.4324/9781315608730 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ada90>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "PDF déjà téléchargé pour E. Miele et al., ICTON 2020, 978-1-8423 2020 IEEE\n",
      "PDF déjà téléchargé pour Stephen et al. Proc. SPIE 6530, 2007\n",
      "PDF déjà téléchargé pour J. Huang et al. (2021)\n",
      "PDF déjà téléchargé pour M. McSherry et al. Sensor review Vol. 25 (2005)\n",
      "PDF déjà téléchargé pour M. Faraday. The Bakerian Lecture: Experimental Relations of Gold (and Other Metals), Light Philos. Trans. R. Soc. London. 1857. 147. 145-181\n",
      "PDF déjà téléchargé pour Wadell et al., ACS Nano, 2014, 8, 12, 11925-11940.\n",
      "PDF déjà téléchargé pour Willets et al Annu Rev Phys Chem 2007 58 267.07\n",
      "PDF déjà téléchargé pour Stewart et al., Chem. Rev., 2008, 108, 494-521.\n",
      "PDF déjà téléchargé pour Willets et al., Annu. Rev. Phys. Chem., 2007, 58, 267-97.\n",
      "PDF déjà téléchargé pour J. N. Anker et al., Nature Materials, 2008. 7.\n",
      "Téléchargement du PDF pour P.R. West et al. Laser and photonic review, 2020\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.1117/1.jbo.25.4.040902 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x113557c50>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.1117/1.jbo.25.4.040902 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x114287610>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.1117/1.jbo.25.4.040902 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ada90>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.1117/1.jbo.25.4.040902 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ad590>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "PDF déjà téléchargé pour Willets et al., Annu. Rev. Phys. Chem., 2007, 58, 267-97.\n",
      "PDF déjà téléchargé pour M. L. Brongersma, Faraday Discussions, 2015, 178, 9-36\n",
      "PDF déjà téléchargé pour Rifat et al., Sensors and Actuators B, 2017, 243, 311-325.\n",
      "PDF déjà téléchargé pour E. Larsson et al., Nanophotonics, 2012, 1, 249-266.\n",
      "PDF déjà téléchargé pour J. Dai et al.. Sensors. 2017. 17. 577. Wadell et al., ACS Nano, 2014, 8, 12, 11925-11940.\n",
      "PDF déjà téléchargé pour E.M. Larson et al. Nano Lett. Vol.7, N°5 (2007)\n",
      "PDF déjà téléchargé pour I. Zoric, E.M Larsson, Advanced Materials, 22, 2010\n",
      "PDF déjà téléchargé pour E.M. Laarson et el. Science. Vo. 326 (2009)\n",
      "PDF déjà téléchargé pour C. Langhammer et al. Nano Lett., 10, 2010\n",
      "PDF déjà téléchargé pour C. Langhammer et al. Nano. Lett., 10, 2010\n",
      "PDF déjà téléchargé pour L. Tian, Nanotechnology 23(2012)\n",
      "PDF déjà téléchargé pour V. Gusak et al. NANO letters, 12 (2012)\n",
      "PDF déjà téléchargé pour Y. Jin et al. PNAS Vol. 115, 544, 2008\n",
      "Téléchargement du PDF pour L. Albero Blanquer, MESC Master thesis, 2018.\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.25148/mmeurs.010558 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ad090>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.25148/mmeurs.010558 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142aca50>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.25148/mmeurs.010558 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ac690>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.25148/mmeurs.010558 (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ae210>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "PDF déjà téléchargé pour T. Guo et al.. Sensors 2017 17. 2732.\n",
      "PDF déjà téléchargé pour T. Guo et al., Jouranl of Lightwave Tech., 2017, 35, 16\n",
      "Téléchargement du PDF pour J.Lao et al., Light: Science & Applications, 2018, 7, 34.\n",
      "Trying with domain: https://sci-hub.se\n",
      "Error accessing https://sci-hub.se: HTTPSConnectionPool(host='sci-hub.se', port=443): Max retries exceeded with url: /10.1038/s41377-018-0040-y (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142ae710>: Failed to resolve 'sci-hub.se' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.st\n",
      "Error accessing https://sci-hub.st: HTTPSConnectionPool(host='sci-hub.st', port=443): Max retries exceeded with url: /10.1038/s41377-018-0040-y (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142aec10>: Failed to resolve 'sci-hub.st' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ru\n",
      "Error accessing https://sci-hub.ru: HTTPSConnectionPool(host='sci-hub.ru', port=443): Max retries exceeded with url: /10.1038/s41377-018-0040-y (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142af110>: Failed to resolve 'sci-hub.ru' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Trying with domain: https://sci-hub.ee\n",
      "Error accessing https://sci-hub.ee: HTTPSConnectionPool(host='sci-hub.ee', port=443): Max retries exceeded with url: /10.1038/s41377-018-0040-y (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x1142af610>: Failed to resolve 'sci-hub.ee' ([Errno 8] nodename nor servname provided, or not known)\"))\n",
      "Failed to locate PDF on all Sci-Hub domains\n",
      "PDF déjà téléchargé pour S.A. Pradanawati et al., Jouml of power source (2016)\n",
      "PDF déjà téléchargé pour Lemaire et al.. Anal. Chem. 2020.\n",
      "PDF déjà téléchargé pour T. Ohzuku et al. J. Electrochem. Soc. (1997)\n",
      "PDF déjà téléchargé pour N. Kircheva et al. J. Electrochem. Soc. (2011)\n",
      "PDF déjà téléchargé pour K. Rhodes et al. Journal of the electrochemical society. 157 (12) 2010\n",
      "PDF déjà téléchargé pour K. Rhodes et al. Review of Scientific Instrumentation (2011)\n",
      "PDF déjà téléchargé pour Chan-yang Choe et al. Materials transactions (2)(2015)\n",
      "PDF déjà téléchargé pour Z. Deng, et al. Joule 4, 1-13 (2020)\n",
      "PDF déjà téléchargé pour A.G. Hsieh et al. Energy Environ. Sci; (2015) DOI:10.1039/c5ee0011k\n",
      "PDF déjà téléchargé pour AG Hsieh et al., Energy Environ. Sci. 8, 1569-1577 (2015).\n",
      "PDF déjà téléchargé pour L Gold et al.. J. Power Sources. 343 536-544 (2017).\n",
      "PDF déjà téléchargé pour P Ladpli, F Kopsaftopoulos, F-K Chang, J. Power Sources. 384, 342-354 (2018).\n"
     ]
    }
   ],
   "source": [
    "# Open references_with_doi_clean.csv\n",
    "df = pd.read_csv(\n",
    "    \"../4_references_to_doi/references_with_doi_clean.csv\",\n",
    "    sep=\";\",\n",
    "    engine=\"python\",\n",
    "    quotechar='\"'\n",
    ")\n",
    "\n",
    "destination_folder = \"Articles cours sensing JMT\"\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    # Check if pdf is already download\n",
    "    if os.path.exists(f\"{destination_folder}/{row['ID']}.pdf\"):\n",
    "        print(f\"PDF déjà téléchargé pour {row['Reference']}\")\n",
    "        continue\n",
    "        \n",
    "    # Download PDF\n",
    "    print(f\"Téléchargement du PDF pour {row['Reference']}\")\n",
    "    get_pdf_from_doi_url(row['DOI'], destination_folder, row['ID'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
